x-pack/plugins/stack_connectors/server/connector_types/openai/openai.ts

```diff
@@ -194,24 +237,64 @@ export class OpenAIConnector extends SubActionConnector<Config, Secrets> {
   }
 
   /**
+   * Streamed security solution AI Assistant requests (non-langchain)
    * Responsible for invoking the streamApi method with the provided body and
-   * stream parameters set to true. It then returns a Transform stream that processes
-   * the response from the streamApi method and returns the response string alone.
+   * stream parameters set to true. It then returns a ReadableStream, meant to be
+   * returned directly to the client for streaming
    * @param body - the OpenAI Invoke request body
    */
   public async invokeStream(body: InvokeAIActionParams): Promise<PassThrough> {
+    const { signal, ...rest } = body;
+
     const res = (await this.streamApi({
-      body: JSON.stringify(body),
+      body: JSON.stringify(rest),
       stream: true,
+      signal,
     })) as unknown as IncomingMessage;
 
     return res.pipe(new PassThrough());
   }
 
   /**
-   * Deprecated. Use invokeStream instead.
-   * TODO: remove once streaming work is implemented in langchain mode for security solution
-   * tracked here: https://github.com/elastic/security-team/issues/7363
+   * Streamed security solution AI Assistant requests (langchain)
+   * Uses the official OpenAI Node library, which handles Server-sent events for you.
+   * @param body - the OpenAI Invoke request body
+   * @returns a Stream<ChatCompletionChunk> array "teed", teed[0] is used for the UI and teed[1] for token tracking
+   * the result is meant to be read/transformed on the server and sent to the client via Server Sent Events
+   */
+  public async invokeAsyncIterator(
+    body: InvokeAIActionParams
+  ): Promise<Array<Stream<ChatCompletionChunk>>> {
+    try {
+      const { signal, ...rest } = body;
+      const messages = rest.messages as unknown as ChatCompletionMessageParam[];
+      const requestBody: ChatCompletionCreateParamsStreaming = {
+        ...rest,
+        stream: true,
+        messages,
+        model:
+          rest.model ??
+          ('defaultModel' in this.config ? this.config.defaultModel : DEFAULT_OPENAI_MODEL),
+      };
+      const stream = await this.openAI.chat.completions.create(requestBody, {
+        signal,
+      });
+      // splits the stream in two, teed[0] is used for the UI and teed[1] for token tracking
+      const teed = stream.tee();
+      return teed;
+      // since we do not use the sub action connector request method, we need to do our own error handling
```

Comment:
Ok, I remember now this was the problem and I had to do it at the connector level. If I don't tee it off the beginning and try to work with the stream in multiple places:

`Cannot iterate over a consumed stream, use .tee() to split the stream.`

Reply:
Yah, feels like this function shouldn't return the `teed` version, just the stream.  Let the caller's tee it if neccessary.

Reply:
if the callers try to tee and another caller has already consumed the stream (even by calling tee), it results in this error.

Reply:
I'm not sure if I follow, there should be only one consumer right?

Reply:
No, LangChain and token counter. if i try to tee it at either, I get that error. If I try to interact with the stream at all and it has already been used, I get that error. Notice that Stream is imported from `openai/streaming`, not from the node package.

I got to this strategy by following the advice of an OpenAI Leader: 

> Stream Duplication: If you need to read from a stream multiple times, use the .tee() method right after the creation of the stream to duplicate it. 
> https://community.openai.com/t/error-cannot-iterate-over-a-consumed-stream/475898

Reply:
Is this because you still return the original stream to the action consumer? I guess you can tee off of the original stream in `executeHelper`, use one of the streams for token tracking, and return the other one to the action consumer.

Reply:
`executeHelper` runs after the executor result has been delivered. It would be too late to tee, as the result will be used first, or perhaps the `executeHelper` would tee first and then the result stream would be broken

Reply:
Is the problem that the `tee` needs to be made ASAP after the original stream is created?  It could be "consumed" somehow by OpenAI code before it actually gets back to the caller?

I guess my worry is this.  We may end up having code wanting to use this, but only needs one stream.  But they're getting two back.  What are the semantics?  Do I need to close the one I'm not using, in case it could block the one I'm actually reading somehow?  Or do I need to read from both, and throw away the result from the one I don't care about?  I don't want to put caller's in a position to have to do this.

Maybe it always makes sense to have a LangChain AND token counter stream?  (clearly showing my n00biness here).  If so, we should probably return a structured `{langChainStream: <>, tokenCountStream: <>}`.

If we think there may be cases where some caller may want a single stream, we could indicate the number of streams we want returned as an optional param, and always return an array of them.

Reply:
We will always use both streams, the consumer will use the first stream and the token tracking will be done on the second stream. Every request needs token tracking, so every request will have 2 streams. Definitely see the problem with confusion about which stream you're supposed to use. If the consumer tries to use the second stream, they will get the tee error. But I'm also ok with naming it for clarity `{consumerStream: <>, tokenCountStream: <>}`. Maybe `consumer` because the consumer could be a tool like LangChain, or it could be in the UI itself. What do you think?

Reply:
returning named streams makes sense to me, given ^^^

Reply:
AFAICT, executeHelper executes the action: https://github.com/elastic/kibana/blob/fb3b66256e93e116e61687565c5b262ad57c8a88/x-pack/plugins/actions/server/lib/action_executor.ts#L491.

it also starts the token tracking: https://github.com/elastic/kibana/blob/fb3b66256e93e116e61687565c5b262ad57c8a88/x-pack/plugins/actions/server/lib/action_executor.ts#L574.

Maybe I'm misunderstanding, but it seems like what you can do here is to call tee(), and modify `rawResult` to return one of the teed streams, and use the other one for token tracking. That way the consumer doesn't have to deal with two streams.

Reply:
Ah ok, I think I was misunderstanding. Still, I think I like the solution better to return named streams from the connector and try to not introduce more connector specific code into the actions executor. Here is that commit https://github.com/elastic/kibana/pull/174126/commits/6ca88fe54a07b8fe2b6b6d1f3e4ac85ddbf68366